File: engine/src/main/java/com/bartlomiejpluta/base/core/world/tileset/model/TileSet.java
package com.bartlomiejpluta.base.core.world.tileset.model;

import com.bartlomiejpluta.base.core.gl.object.material.Material;
import com.bartlomiejpluta.base.core.gl.object.mesh.Mesh;
import com.bartlomiejpluta.base.core.gl.object.texture.Texture;
import lombok.Getter;

public class TileSet {
   private final Texture texture;
   private final int rows;
   private final int columns;
   private final float columnStep;
   private final float rowStep;
   private final Mesh mesh;

   @Getter
   private final int tileWidth;

   @Getter
   private final int tileHeight;

   public TileSet(Mesh mesh, Texture texture, int rows, int columns, int tileWidth, int tileHeight) {
      this.texture = texture;
      this.rows = rows;
      this.columns = columns;
      this.columnStep = 1/(float) columns;
      this.rowStep = 1/(float) rows;
      this.tileWidth = tileWidth;
      this.tileHeight = tileHeight;
      this.mesh = mesh;
   }

   public Tile getTile(int row, int column) {
      var material = Material.textured(texture);
      material.setSpriteSize(columnStep, rowStep);
      material.setSpritePosition(column * columnStep, row * rowStep);
      return new Tile(mesh, material, tileWidth, tileHeight);
   }
}

File: editor/src/main/kotlin/com/bartlomiejpluta/base/editor/model/tileset/TileSet.kt
package com.bartlomiejpluta.base.editor.model.tileset

import com.bartlomiejpluta.base.editor.model.map.brush.Brush
import javafx.scene.image.Image
import javafx.scene.image.PixelFormat
import javafx.scene.image.WritableImage
import java.nio.ByteBuffer


class TileSet(private val image: Image, val rows: Int, val columns: Int) {
   val tileWidth = image.width.toInt() / columns
   val tileHeight = image.height.toInt() / rows
   val width = tileWidth * columns
   val height = tileHeight * rows

   val tiles: Array<Array<Tile>> =
      Array(rows) { row -> Array(columns) { column -> cropTile(row, column) } }

   val baseBrush: Brush
      get() = Brush(arrayOf(arrayOf(tiles[0][0])))

   private fun cropTile(row: Int, column: Int): Tile {
      val reader = image.pixelReader
      val buffer = ByteBuffer.allocate(tileWidth * tileHeight * 4)
      reader.getPixels(
         column * tileWidth,
         row * tileHeight,
         tileWidth,
         tileHeight,
         PixelFormat.getByteBgraInstance(),
         buffer,
         4 * tileWidth
      )
      val tile = WritableImage(tileWidth, tileHeight)
      val writer = tile.pixelWriter
      writer.setPixels(0, 0, tileWidth, tileHeight, PixelFormat.getByteBgraInstance(), buffer, 4 * tileWidth)

      return Tile(this, row, column, tile)
   }

   fun getTile(row: Int, column: Int) = tiles[row.coerceIn(0 until rows)][column.coerceIn(0 until columns)]

   fun getTile(id: Int) = tiles[id / columns][id % columns]
}